workflow/controller/operator.go

```diff
@@ -1719,6 +1729,21 @@ func getChildNodeIndex(node *wfv1.NodeStatus, nodes wfv1.Nodes, index int) *wfv1
 	return &lastChildNode
 }
 
+func getChildNodeIdsRetried(node *wfv1.NodeStatus, nodes wfv1.Nodes) []string {
+	childrenIds := []string{}
+	r := regexp.MustCompile(`^` + regexp.QuoteMeta(node.Name) + `\(\d+\)`)
+	for i := 0; i < len(node.Children); i++ {
+		n := getChildNodeIndex(node, nodes, i)
+		if node == nil {
+			continue
+		}
+		if r.MatchString(n.Name) {
+			childrenIds = append(childrenIds, n.ID)
+		}
+	}
+	return childrenIds
```

Comment:
Can you add comments to this function? I'm trying to understand why you're not just looking for all child nodes of "Type"="Retry"

Reply:
As commented, we can't distinguish retried nodes from other nodes because they have the same node type `type: Pod`.
Only the parent node has `type: Retry`.
(I also don't like this function I wrote, so any suggestions are very welcome.)

example parent node

```
    lovely-bear:
      id: lovely-bear
      name: lovely-bear
      displayName: lovely-bear
      type: Retry
      templateName: argosay
      templateScope: local/lovely-bear
      phase: Failed
      message: No more retries left
      startedAt: '2023-09-30T13:49:20Z'
      finishedAt: '2023-09-30T13:49:56Z'
      progress: 2/6
      resourcesDuration:
        cpu: 42
        memory: 30
      outputs:
        artifacts:
          - name: main-logs
            s3:
              key: lovely-bear/lovely-bear-argosay-74551197/main.log
        exitCode: '1'
      children:
        - lovely-bear-477066958
        - lovely-bear-794339884
        - lovely-bear-4167892875
        - lovely-bear-410250672
        - lovely-bear-74551197
        - lovely-bear-2163413524
```

example child node

```
    lovely-bear-410250672:
      id: lovely-bear-410250672
      name: lovely-bear(2)
      displayName: lovely-bear(2)
      type: Pod
      templateName: argosay
      templateScope: local/lovely-bear
      phase: Failed
      message: Error (exit code 1)
      startedAt: '2023-09-30T13:49:38Z'
      finishedAt: '2023-09-30T13:49:43Z'
      progress: 0/1
      resourcesDuration:
        cpu: 7
        memory: 5
      outputs:
        artifacts:
          - name: main-logs
            s3:
              key: lovely-bear/lovely-bear-argosay-410250672/main.log
        exitCode: '1'
      hostNodeName: k3d-k3s-default-server-0
```

example hooked node

```
    lovely-bear-2163413524:
      id: lovely-bear-2163413524
      name: lovely-bear.hooks.failed
      displayName: lovely-bear.hooks.failed
      type: Pod
      templateName: hook
      templateScope: local/lovely-bear
      phase: Succeeded
      startedAt: '2023-09-30T13:49:56Z'
      finishedAt: '2023-09-30T13:50:01Z'
      progress: 1/1
      resourcesDuration:
        cpu: 7
        memory: 5
      outputs:
        artifacts:
          - name: main-logs
            s3:
              key: lovely-bear/lovely-bear-hook-2163413524/main.log
        exitCode: '0'
      hostNodeName: k3d-k3s-default-server-0
```

Reply:
Yes, thank you! I forgot the parent was "Retry" and the children were "Pod"

Reply:
Okay, starting to understand the logic now looking at what you posted. Thanks for doing that.

So, I can assume that the spec for this is a running hook which has `expression: workflow.status == "Running"`, right? Does it seem a little odd that the hook node is a child of the Retry node? Do you figure that it would be the child of whatever the first running Node is?

Reply:
I also can't really remember what is a child of what. If there are 3 Steps A-> B-> C, do you know if B would be a child of A because it follows it, or are A, B, and C siblings (which is what I would assume)?

Reply:
> So, I can assume that the spec for this is a running hook which has expression: workflow.status == "Running", right?

Yes

> Does it seem a little odd that the hook node is a child of the Retry node?

I think, not.

1. Hooks should be triggered by retry node status.
   * `Succeed`/`Failed` hook should be triggered when `Retry` node succeeded/failed.
   * So `Running` hook should be triggered depending on `Retry` node status, shouldn't be?
2. It might be difficult technically.
   * We should change hook logic around [here](https://github.com/argoproj/argo-workflows/blob/26481a2146107ad0937ef7698c27f3686f93c81e/workflow/controller/steps.go#L316-L317)

> I also can't really remember what is a child of what. If there are 3 Steps A-> B-> C, do you know if B would be a child of A because it follows it, or are A, B, and C siblings (which is what I would assume)?

I don't fully understand what you mentioning here, but it depends on steps

a: parent -> b, c: childs

```yaml
 - - name: a
 - - name: b
    - name: c
``` 

a, b, c: parents

```yaml
 - - name: a
   - name: b
   - name: c
```

Reply:
Thanks for clarifying. I see now that the relationship from parent to child is "child follows parent sequentially", right? Since the hook is on the workflow level, I do have to wonder what it would look like if there were 2 steps running in parallel and the same hook of expression `workflow.status == "Running"`. What would the hook be a child of? Whichever parallel step happened to get started first?

Reply:
In the example from your Issue, the Hook is on the [workflow](https://github.com/argoproj/argo-workflows/blob/master/examples/life-cycle-hooks-wf-level.yaml) level, not the [step](https://github.com/argoproj/argo-workflows/blob/master/examples/life-cycle-hooks-tmpl-level.yaml) level, right?

Reply:
Thank you for pointing that out!
When we use step with WorkflowLevelHook, a hook is connected to the Steps node, so retryStrategy works fine.
![image](https://github.com/argoproj/argo-workflows/assets/83329336/75eed3a3-3915-4f3b-810e-953c72667230)

<details>
<summary>workflow</summary>

```yaml
spec:
  templates:
    - name: hello-hello-hello
      inputs: {}
      outputs: {}
      metadata: {}
      steps:
        - - name: a
            template: whalesay
            arguments: {}
          - name: b
            template: whalesay
            arguments: {}
    - name: whalesay
      inputs: {}
      outputs: {}
      metadata: {}
      container:
        name: ''
        image: docker/whalesay
        command:
          - sh -c
        args:
          - exit 1
        resources: {}
      retryStrategy:
        limit: '1'
    - name: hook
      inputs: {}
      outputs: {}
      metadata: {}
      container:
        name: ''
        image: docker/whalesay
        command:
          - cowsay
        args:
          - test
        resources: {}
      retryStrategy:
        limit: '0'
  entrypoint: hello-hello-hello
  arguments: {}
  hooks:
    failed:
      template: hook
      arguments: {}
      expression: workflow.status == "Failed"
    running:
      template: hook
      arguments: {}
      expression: workflow.status == "Running"
    succeed:
      template: hook
      arguments: {}
      expression: workflow.status == "Succeeded"
```

</details>

But we have the same problem for [WorkflowTemplateLevelHook](https://github.com/argoproj/argo-workflows/issues/11589#issuecomment-1743902549)
I added validation for WorkflowTemplateLevelHook in the PR description

Reply:
I see. So, this PR resolves both cases? And is this particular method - `getChildNodeIdsRetried()` - the logic that runs in both cases? You were saying the hook in the Workflow-level case is a child of the StepGroup?

Reply:
I do see what you mean that going off the regular expression of the name doesn't seem great. It does make me wonder if perhaps the NodeStatus could potentially have some new key/value pair in it or potentially even the "Type" field could have a new value like "HookPod" - what are your thoughts since you've looked at this code more than I?

Reply:
I am kind of thinking we shouldn't rely on somebody not changing the name in the future

Reply:
> I see. So, this PR resolves both cases? And is this particular method - getChildNodeIdsRetried() - the logic that runs in both cases?

Yes. This PR resolves both cases.

> It does make me wonder if perhaps the NodeStatus could potentially have some new key/value pair in it like isHook - what do you think?

Ideally, completely agree with you.
I'm going to write a code that changes NodeStatus go struct!

Reply:
thank you! Once you do I think you'll need to run `make codegen` to auto-generate the CRD and other files from that struct

Reply:
Hi, I've added `NodeStatus.Hooked` and some node initialization. Could you give me feedback :)?

Reply:
can you add a thorough comment to the top of this function to explain what it's doing, what the various cases are, etc?